lib/codegen/src/constant_folding.rs

```diff
@@ -0,0 +1,240 @@
+            Some(ConstImm::Ieee64(imm_as_f64))
+        }
+        UnaryBool {
+            opcode: Bconst,
+            imm,
+        } => Some(ConstImm::Bool(imm)),
+        _ => None,
+    }
+}
+
+fn evaluate_numerical_binary(
+    dfg: &ir::DataFlowGraph,
+    opcode: ir::Opcode,
+    args: [ir::Value; 2],
+) -> Option<ConstImm> {
+    use std::num::Wrapping;
+
+    let (imm0, imm1) = (
+        resolve_value_to_imm(dfg, args[0])?,
+        resolve_value_to_imm(dfg, args[1])?,
+    );
+
+    match opcode {
+        ir::Opcode::Iadd => {
+            let imm0 = Wrapping(imm0.unwrap_i64());
+            let imm1 = Wrapping(imm1.unwrap_i64());
+            Some(ConstImm::I64((imm0 + imm1).0))
+        }
+        ir::Opcode::Isub => {
+            let imm0 = Wrapping(imm0.unwrap_i64());
+            let imm1 = Wrapping(imm1.unwrap_i64());
+            Some(ConstImm::I64((imm0 - imm1).0))
+        }
+        ir::Opcode::Imul => {
+            let imm0 = Wrapping(imm0.unwrap_i64());
+            let imm1 = Wrapping(imm1.unwrap_i64());
+            Some(ConstImm::I64((imm0 * imm1).0))
+        }
+        ir::Opcode::Udiv => {
+            let imm0 = Wrapping(imm0.unwrap_i64());
+            let imm1 = Wrapping(imm1.unwrap_i64());
+            if imm1.0 == 0 {
+                panic!("Cannot divide by a zero.")
+            }
+            Some(ConstImm::I64((imm0 / imm1).0))
+        }
+        ir::Opcode::Fadd => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => Some(ConstImm::Ieee32(imm0 + imm1)),
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => Some(ConstImm::Ieee64(imm0 + imm1)),
+            _ => unreachable!(),
+        },
+        ir::Opcode::Fsub => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => Some(ConstImm::Ieee32(imm0 - imm1)),
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => Some(ConstImm::Ieee64(imm0 - imm1)),
+            _ => unreachable!(),
+        },
+        ir::Opcode::Fmul => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => Some(ConstImm::Ieee32(imm0 * imm1)),
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => Some(ConstImm::Ieee64(imm0 * imm1)),
+            _ => unreachable!(),
+        },
+        ir::Opcode::Fdiv => match (imm0, imm1) {
+            (ConstImm::Ieee32(imm0), ConstImm::Ieee32(imm1)) => {
+                if imm1 == 0.0 {
+                    panic!("Cannot divide by a zero.")
+                }
+                Some(ConstImm::Ieee32(imm0 / imm1))
+            }
+            (ConstImm::Ieee64(imm0), ConstImm::Ieee64(imm1)) => {
+                if imm1 == 0.0 {
+                    panic!("Cannot divide by a zero.")
+                }
+                Some(ConstImm::Ieee64(imm0 / imm1))
+            }
+            _ => unreachable!(),
+        },
+        // ir::Opcode::Fadd => Some(imm0.unwrap)
+        _ => None,
+    }
```

Comment:
Yes, this looks like a good overall direction. To further separate the arithmetic from the IR walking, we can pull just this match statement out into a separate function in a separate file.

Reply:
[preopt](https://github.com/CraneStation/cranelift/blob/master/lib/codegen/src/preopt.rs#L447) does this. So it's not something we need here right away. However if this constant folding pass starts growing into a more general optimization framework, it may want to start doing that kind of thing too, at which point we should think about how to organize things.